### ISSUE TITLE
Make the modules semantic and easy to use

### ISSUE THREAD

**Issue by alexewerlof** [schibsted/eslint-config-spt#59]
Re #56,

I've looked into [airbnb config](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb-base) and it's a bit unnecessarily too complicated. Hold your 🐴 es. Let me explain:

They configure their parser [for ES2017](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/index.js#L12) which is obviously a good choice if you transpile your code (and most people in front-end land do) but we have a good deal of Node projects that contain code that doesn't need compilation.

There are other things as well that I recommend you to read if you're curious but I want to propose a simple *usage based*  config structure that groups rules together based on how javascript is used in a project.

* **base.js:** contains only ES5 rules regardless of runtime environment. This is like [airbnb style](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js), [best practices](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js) and [variable usage](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/variables.js) rules combined.
* **node-lts.js:** contains configs that are understandable by node LTS (aka v6.x). It builds on top of **base.js**. This is like [airbnb node config](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/node.js).
* **node-current.js:** same as **node-lts.js** but for projects that use the latest version of node. This will build on top of **node-lts** but will add/modify some settings.
* **browser-legacy.js:** for projects that are supposed to run in the browser without transpilation (like [SPiD's SDK.js](https://github.com/schibsted/sdk-js)). This may be slightly different from **base.js**
* **browser-latest.js** just joking! 😄 There's no such a thing for obvious reasons! But for those of us who transpile latest ES17, we need something equivalent to what airbnb provides by default.
* more... we probably have more use-cases for how we use this versatile language (for example we have some internally used chrome extensions or some may use react-native), but these are use-cases that I've faced.

The idea is that the users should be able to include a rule set individually with minimum side effects from other runtimes (browser vs node) that they don't use. It's an extension of what is already implemented in `eslint-config-schibsted`:

![screen shot 2017-01-04 at 17 40 41](https://cloud.githubusercontent.com/assets/835750/21650385/15487536-d2a5-11e6-9ec9-d226662925eb.PNG)

Now that we use monorepo structure, we may even publish these *chunks* of rules as separate modules? Maybe a crazy idea. Maybe not. I'd love to hear some discussion.

This is just a suggestion 🌻 

Ping: @nosilleg @ivarconr @SimenB @marek-nogiec @trygve-lie

**Comment by SimenB** [comment]
> They configure their parser for ES2017 which is obviously a good choice if you transpile your code (and most people in front-end land do) but we have a good deal of Node projects that contain code that doesn't need compilation.

That's just the parser, it has nothing to do with what rules it allows or disallows. So IMO this should always be latest (and I argued for a `latest` option to eslint before eslint/eslint#7528).

> Now that we use monorepo structure, we may even publish these chunks of rules as separate modules? Maybe a crazy idea. Maybe not. I'd love to hear some discussion.

I think that makes a lot of sense. Then we can publish separate configs with e.g. `eslint-plugin-shibcsted-node` included if we want extra node linting.
Splitting up by target makes sense as well, and your suggestions sound to me 😄

**Comment by nosilleg** [comment]
Typing on phone with sketchy connection, so just getting this out quickly...

I'm all for extending the config options. Be it different packages or the same doesn't matter much, it's just a matter of putting the right one in your devices.

As for names; there are multiple node lts versions. So I don't think it makes sense from a user or repo point of view to have something called just "lts".  Better name TBD, "lts4", "ltsA", ltsArgon" etc.

Will base forever be ES5? Why not have base be `eslint-config-schibsted-es3`, instead. Then have `-es5`, `-es2015`, `-es2016`, `-eslevel4` ... `-eslevel0`.  With each one inheriting from the previous. Similar to how the Babel levels work. 

Also note the `-false` modules which are there because of the directory inheritance in eslint and projects which may have browser code lower than node code, etc. 

Will comment more when I get the chance.

**Comment by userpixel** [comment]
good points @nosilleg . Waiting to read more.
BTW, thanks for taking the time to share your 💭. I heard you're on leave. Your contribution is very much appreciated 👍

**Comment by nosilleg** [comment]
Ok.  More feedback.  The above was opinions on how to improve the suggested changes.  The below is opinions on why we shouldn't do the suggested changes and instead take a different approach.

So the usage based approach is for people that want to use "release version" of JavaScript.  So the eng says "I want to use ES2015", so imports out es2015 ruleset and writes ES2015 code.

A few "issues":

There is no rule for making sure that `Array.prototype.includes` isn't used.  so our ES2015 config will need to specify the `parserOptions` of `ecmaVersion` 6 in order for it to be validated.

There was a 2 week period where Chrome was the only browser that supported the feature.  Firefox followed shortly after.  Edge was around 17 months later. (best guess)  Internet Expolorer will never have the feature.

Looking into the future I don't think it's unreasonable to assume that this will continue to happen.

This kind of breaks the use case of choosing a JavaScript version and writing code for it and hoping that it works on all browsers; unless it's a really old version of JavaScript that already has full browser support.  The `parserOptions` support an es8 setting; what's the browser support like for that?

As such, my preferred workflow would be to use the level4 config (level 4 is the "not standard yet, but unlikely to change" syntax) and transpile it with `babel-preset-env`.

This means you're writing code that eventually every maintained browser will be able to execute, and the size of dist will shrink over time, even if the src is never updated.

The downside to that is the requirement to transpile.  This means some extra dev dependencies and an extra line or two in the deploy scripts.

I think that tradeoff is worthwhile in order to know that the code you are writing will run on all the browsers that you want it to without having to look up every syntax feature as you type it.

Having said that, my desired workflow will work even with the JavaScript versioned based approach, so I'm not impacted much in usage.  I do think there will be an impact in maintenance of this project, tho. 

And yeah, still on leave to the end of the month, so only getting limited time to jump on here.  Apologies if my rushing makes my comments less friendly than they could be.

**Comment by alexewerlof** [comment]
> There was a 2 week period where Chrome was the only browser that supported the feature. Firefox followed shortly after. Edge was around 17 months later. (best guess) Internet Expolorer will never have the feature.

I would argue those browsers were not really ES5/6/7 compatible so the developer shouldn't target them with `eslint-config-schibsted/browser-es6` (or use transpilation).

> As such, my preferred workflow would be to use the level4 config (level 4 is the "not standard yet, but unlikely to change" syntax) and transpile it with babel-preset-env.

I'm mainly working with Node without transpilation so this approach sounds like an unnecessary complication in my workflow.

But you had good points that made me revise the original structure like this:

* **javascript.js:** contains only ES5 rules regardless of runtime environment. This is like [airbnb style](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js), [best practices](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/best-practices.js) and [variable usage](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/variables.js) rules combined. `"parserOptions": { "ecmaVersion": 5,...}`
* **node/base.js:** contains configs that are understandable by node LTS (e.g. v6.x for now). It builds on top of **javascript.js**. This is like [airbnb node config](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/node.js). `"parserOptions": { "ecmaVersion": 6, "sourceType": "module", ...}`
* **node/7.js:** inherits **node/base.js** 👆 but adds stuff for features that were added in 7. Then we'll have **node/8.js**, etc.
* **browser/base.js:** for projects that are supposed to run in the browser without transpilation (like [SPiD's SDK.js](https://github.com/schibsted/sdk-js)). This may be slightly different from **javascript.js** `"parserOptions": { "ecmaVersion": 4, "sourceType": "script", ...}`
* **browser/es5.js** if your target runtime environment is es5 capable browsers that are pretty much in wide use but you don't transpile.
* **browser/latest.js** Will be exactly what @nosilleg proposed 👆 (transpilation, `babel-preset-env`, etc.)

**Comment by nosilleg** [comment]
> I would argue those browsers were not really ES5/6/7 compatible so the developer shouldn't target them with eslint-config-schibsted/browser-es6 (or use transpilation).

At what point did those browsers become ES/5/6/7 compatible?  Are they now? How long will it take you to look it up?  Will you miss any features and implement code that isn't currently supported?  What about for the project that you jump on for three weeks to help another team with?

My opinion is that answering those quesstions and maintaining project code, and this projects configs is a more time consuming task than it is to add in a transpilation step to the project.

I would personally like to see Schibsted adopt a common coding standard, not a collection of coding standards.

This is much more inline with how the project currently stands, but opens it up to post es6 features. (where there is a base, and then env files for setting environment specific globals and rules, e.g. React, Node, Browser.)

**Comment by SimenB** [comment]
>There is no rule for making sure that Array.prototype.includes isn't used. so our ES2015 config will need to specify the parserOptions of ecmaVersion 6 in order for it to be validated.

There's a difference between supported API and supported syntax. `ecmaVersion` just means what eslint is able to _parse_, not what it reports errors on (too low level makes eslint itself throw on `const`, if your target environment doesn't support `const`, that should be handled by the `no-var` rule, it should still parse the code).

I don't think it's the linter's responsibility to tell you when you use an API which may not be available in the browser (makes more sense for node (can use https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-unsupported-features.md for that)), that's up to the developer. At FINN we have a common polyfill that all webapps include (https://npm.im/@finn-no/polyfills).

---

I think it's fair to make the default config target es2015. We can have something called legacy which is ES5 (or even ES3) compliant without transpilation, but it shouldn't be the default.
I also think it's fair to have lint rules for all LTSes and `latest` node version (meaning 4, 6 and 7 now, very soon 4, 6 and 8). Having `node/base.js` be node@4 for now is fine.

So I think I agree with @userpixel's last proposal, although I think `browser/base.js` and `browser/es5.js` should be combined. If you don't transpile, just target something really low.

**Comment by nosilleg** [comment]
@SimenB 

> it should still parse the code

So we're in agreement on that.  This is contrary to @userpixel's suggestion that ES5 should have ecmaVersion 5.

> I don't think it's the linter's responsibility to tell you when you use an API which may not be available in the browser ... that's up to the developer.

And that's exactly the folly that my last response was about.  Why put that burden on the developer?  It should be the linters job to tell the user that their syntax isn't correct, even if that syntax is supported by some other/possible future version.

> At FINN we have a common polyfill that all webapps include (https://npm.im/@finn-no/polyfills).

And `babel-preset-env` will provide that polyfills for you as well, without the dev having the think about which ones are required and which ones are not.

Again removing burden from the developer in exchange for a CI config tweak.

> I also think it's fair to have lint rules for all LTSes and latest node version (meaning 4, 6 and 7 now, very soon 4, 6 and 8).

`babel-preset-env` takes care of this as well.  

Why write to many standards, when you can write to one and the code works where you want it to work?

**Comment by SimenB** [comment]
>And that's exactly the folly that my last response was about. Why put that burden on the developer? It should be the linters job to tell the user that their syntax isn't correct, even if that syntax is supported by some other/possible future version.

It's not syntax though, that's my point. There's no way a linter such as ESLint can now what methods are available and which are not. Maybe the developer has done `Array.prototype.contains = function () {return Array.prototype.includes.apply(this, arguments)}`, how would eslint know that `['foo', 'bar'].contains('foo')` is valid and won't throw runtime? If you really want this, you should use something like Flow or TypeScript.

If you know of a plugin for ESLint (not babel) that can tell me that the function I'm trying to call is not available in my target env, please link it 😄 

>And babel-preset-env will provide that polyfills for you as well, without the dev having the think about which ones are required and which ones are not.

We have a separate polyfill package for all apps, so that the user isn't forced to download the same polyfill multiple times. www.finn.no is built up from multiple different webapps, and is used a lot on mobile devices, so we save the user quite a lot of bandwidth (and time one older devices) by having the polyfill already downloaded as soon as they hit our site, no matter which part of it is it.

In addition, the code I write at work is almost entirely server-side only, with no babel at all. I agree that`babel-preset-env` is very nice to use if you're going to transpile, but forcing babel on code written for the backend is a no-go, IMO. It's not "a CI config tweak", it's a completely different way of packaging up and sharing your code.

---

That said, except for the fact I think babel on the backend should be optional, we are kinda in agreement, I think. 😀

My suggestions:

- We have a common base for style (I think prettier should mostly handle that part, but there might be a couple of things that doesn't cover) and best practices.
- For the browser, we should have two targets: with and without babel (I can agree on not targeting untranspiled, but if someone wants to maintain the config, why not?).
- Lastly, a config targeting node. Maybe for all LTSes and "current", but I can work with just targeting latest LTS to keep the maintenance burden down.
  - If code is transpiled, it could use the browser config. Maybe add in `eslint-plugin-node` if that's wanted? That might be more in the real of a react config, though 😄

**Comment by nosilleg** [comment]
I think the meaning of my original comment didn't come across correctly. We can't limit to ecmaVersion 5 because that blocks usage of new features that can't be guarded against by (default) linting rules.

But to address your comments...

---

> how would eslint know that `['foo', 'bar'].contains('foo')` is valid and won't throw runtime?

Safe to presume that we're keeping `no-extend-native`?

> If you know of a plugin for ESLint (not babel) that can tell me that the function I'm trying to call is not available in my target env, please link it 😄

How about the reverse? eslint-plugin-unicorn's `prefer-starts-ends-with` rule tells you that you should be using methods which may cause runtime errors if your not polyfilling.

Given the `no-extend-native` it's not a stretch to write a rule the other way that says not to use `startsWith`.

---

I don't think we should have two browser targets. If your not writing to our standard then pick a different standard.

ie Use e-c-s and transpile or use Airbnb legacy if not transpiling.

If someone really has a desire to maintain a legacy config they can PR Airbnb. 
In other words, just because there is a maintainer today doesn't mean there will be one tomorrow.
Why do we want to take on the burden? Why do we want to encourage a split community?

---

Why the dislike for transpile on the server?  Is it just pre runtime transpile that you don't like, or are you not using any hooks like babel-register either?

**Comment by nosilleg** [comment]
Ps typing on phone, so apologies for the lack of pleasantries.  I'm glad the convo has picked up again.

**Comment by alexewerlof** [comment]
@nosilleg oh I was just gonna say: "congratulations for writing the shortest post in this thread" but... :smile:

**Comment by nosilleg** [comment]
@userpixel Maybe my next one.

**Comment by nosilleg** [comment]
Addressed in https://github.com/schibsted/eslint-config-schibsted/pull/74 by going with the One Standard approach.  It does not provide support for legacy syntax.

**Comment by nosilleg** [comment]
Closing as per my last message.
